### DISCUSSION TITLE
AES-GCM in stream mode

### DISCUSSION THREAD

**Discussion by jiep** [paulmillr/noble-ciphers#36] [Q&A]
Hello,

First, I am grateful for your excellent work on this library. The quality and attention to detail are truly impressive and have been a great help in my projects.

I have a question regarding the use of the library. Is it possible to use AES-GCM in stream mode, similar to how it can be done in Node.js? This feature would be handy for my current project, and I'm hoping there's a way to implement it in your library.

For reference, here is an example of how this is achieved in Node.js:

```js
// https://nodejs.org/api/crypto.html#class-cipher
const cipher = crypto.createCipheriv(mode, key, iv);
const outputChunk1 = cipher.update(inputChunk1); // Is stream mode supported on Noble?
const outputChunk2 = cipher.update(inputChunk2);
// ...
const outputChunkN = cipher.update(inputChunkN);
cipher.final();
const tag = cipher.getAuthTag();
```

Thank you again for your hard work and for providing such a fantastic library. I look forward to your response.

Best regards.

**Comment by larabr** [comment]
I'm also interested in a streaming interface, and for other AES modes as well.
For now I I implemented support myself in my code using the exported `unsafe` functions, but I'd rather not have to think about it 😁

**Comment by paulmillr** [comment]
I think the current architecture keeps RAM usage small. So, besides RAM usage, what would be usage for this?

Also, seems complex to implement on top of current architecture. We'd be having an audit ASAP.

**Reply by larabr** [reply]
Not sure if we are on the same page about RAM usage; if we get 1gb of streamed data to process, with the current interface I need to wait for and buffer the whole input to encrypt/decrypt. This is unnecessary for unauthenticated AES modes (eg CBC).

For authenticated modes like GCM, there's a trickier matter when it comes to decryption, as to stream the decrypted data you'd have to release data early (before the tag has been verified). So the function can be used insecurely if the consumer is not aware of this.

> Also, seems complex to implement on top of current architecture. 

Yes, definitely. It's part of the reason why I am not opening a PR myself for this 😓

**Reply by alterstep** [reply]
[AES-GCM is not safe for large inputs](https://libsodium.gitbook.io/doc/secret-key_cryptography/aead#limitations)

**Reply by larabr** [reply]
I don't think this is that relevant when it comes to decide for/against streaming support tbh; it's always possible to throw on encryption for an input above the size limit to add a safeguard, but that doesn't prevent a user from encrypting too many messages with the same key, for instance.

**Reply by paulmillr** [reply]
So, to make a line: it would be nice to get this, but it needs a nice API and such.

I like what you've did in openpgp (https://github.com/openpgpjs/openpgpjs/blob/b2bd8a0fdd12902484d65baa4ae4eb7f146fcd32/src/crypto/cipherMode/cfb.js#L357). It tells ciphers are versatile enough for all kinds of cases. And after the audit, "unsafe" methods would still get audited.

**Reply by ferbs** [reply]
The approach used in that openpgpjs example only works with unauthenticated ciphers. 

> complex to implement on top of current architecture

In terms of minimizing audit impact, would it be feasible to add a utility function that uses the output of an existing encrypt function, discarding its auth tag bytes, updating instead a separate auth tag for each chunk? If so, a TransformStream could handle the rest. (Potentially offering a flag to skip the redundant call to computeTag.)

My use case is encrypting potentially large request bodies and files as they are uploaded to object stores. I'd like the client agent to encrypt the data directly during upload. It will usually not have an s3-client itself, instead using either a pre-signed upload URL or an endpoint proxying to the object store.

**Reply by paulmillr** [reply]
>In terms of minimizing audit impact, would it be feasible to add a utility function that uses the output of an existing encrypt function, discarding its auth tag bytes, updating instead a separate auth tag for each chunk? If so, a TransformStream could handle the rest. (Potentially offering a flag to skip the redundant call to computeTag.)

Maybe. I wonder how that would look in actual code.

**Reply by ferbs** [reply]
I tried to whip up a proof of concept using the lib as-is but it fails with `Error: invalid tag`. Here's the src for the experiment:
https://gist.github.com/ferbs/e018c6a1a1cb312fb4ef2f2de4a11ca1

Is the code in the experiment sloppy/wrong, or would stream support indeed require modifying the lib itself? Prob sloppy code but if the lib, am wondering if a flag to skip its append behavior and repeat-use checks would work, or if it would require something more significant like new *wrapCipher* and *computeTag* utils for streaming.

**Reply by paulmillr** [reply]
Yeah you shouldnt use combined xchacha poly. Instead combination needs to be done in your code — which is easy.
